File: src/IRanges_constructor.c
/****************************************************************************
 *               Support functions for the IRanges constructor              *
 ****************************************************************************/
#include "IRanges.h"

#define	R_INT_MIN (INT_MIN + 1)
#define	R_INT_MAX INT_MAX

static char errmsg_buf[200];


/****************************************************************************
 * solve_user_SEW0()
 */

/* Return -1 if the range specified by 'start', 'end', and 'width' is
   invalid, and 0 otherwise. */
static int solve_range(int start, int end, int width,
		int *solved_start, int *solved_width)
{
	long long int tmp;
	static const char *too_many_NAs = "at least two out of 'start', "
					  "'end', and 'width', must\n  be "
					  "supplied";

	*solved_start = start;
	*solved_width = width;
	if (width == NA_INTEGER) {
		if (start == NA_INTEGER || end == NA_INTEGER) {
			snprintf(errmsg_buf, sizeof(errmsg_buf),
				 "%s", too_many_NAs);
			return -1;
		}
		/* Compute and check 'width'. */
		tmp = (long long int) end - start + 1;
		if (tmp < 0) {
			snprintf(errmsg_buf, sizeof(errmsg_buf),
				 "'end' must be >= 'start' - 1");
			return -1;
		}
		if (tmp > R_INT_MAX) {
			snprintf(errmsg_buf, sizeof(errmsg_buf),
				 "the 'width' (%lld) inferred from the "
				 "supplied 'start'\n  and 'end' is too big "
				 "(>= 2^31)", tmp);
			return -1;
		}
		*solved_width = (int) tmp;
		return 0;
	}
	if (width < 0) {
		snprintf(errmsg_buf, sizeof(errmsg_buf),
			 "negative widths are not allowed");
		return -1;
	}
	if (start == NA_INTEGER) {
		if (end == NA_INTEGER) {
			snprintf(errmsg_buf, sizeof(errmsg_buf),
				 "%s", too_many_NAs);
			return -1;
		}
		/* Compute and check 'start'. */
		tmp = (long long int) end - width + 1;
		if (tmp < R_INT_MIN || tmp > R_INT_MAX) {
			snprintf(errmsg_buf, sizeof(errmsg_buf),
				 "the 'start' (%lld) inferred from the "
				 "supplied 'end'\n  and 'width' is beyond "
				 "the limits of what is currently supported "
				 "(must\n  be > -2^31 and < 2^31 for now)",
				 tmp);
			return -1;
		}
		*solved_start = (int) tmp;
		return 0;
	}
	if (end == NA_INTEGER) {
		if (start == NA_INTEGER) {
			snprintf(errmsg_buf, sizeof(errmsg_buf),
				 "%s", too_many_NAs);
			return -1;
		}
		/* Compute and check 'end'. */
		tmp = (long long int) start + width - 1;
		if (tmp < R_INT_MIN || tmp > R_INT_MAX) {
			snprintf(errmsg_buf, sizeof(errmsg_buf),
				 "the 'end' (%lld) inferred from the "
				 "supplied 'start'\n  and 'width' is beyond "
				 "the limits of what is currently supported "
				 "(must\n  be > -2^31 and < 2^31 for now)",
				 tmp);
			return -1;
		}
		return 0;
	}
	tmp = (long long int) end - start + 1;
	if (width != tmp) {
		snprintf(errmsg_buf, sizeof(errmsg_buf),
			 "the supplied 'width' (%d) doesn't match "
			 "the width\n  inferred from the supplied "
			 "'start' and 'end' (%lld)", width, tmp);
		return -1;
	}
	return 0;
}

/* --- .Call ENTRY POINT ---
  'start' and 'width' can be used **as-is** to construct the IRanges object
  to return if they satisfy at least both criteria:
    (a) They don't have a "dim" or "names" attribute on them.
    (b) They don't contain NAs.
  Note that this just reflects what validObject() expects to see in the
  "start" and "width" slots of an IRanges object.
  If they can't be used **as-is** then they need to be modified (i.e. the
  names need to be removed and/or the NAs in them need to be resolved).
  This requires duplicating them first.
  Of course they also must define valid ranges, that is, after resolution
  of the NAs, the width must be >= 0 and < 2^31, the start must be > -2^31
  and < 2^31, and the implicit end must be > -2^31 and < 2^31. This is
  checked when resolving the NAs.
*/
SEXP solve_user_SEW0(SEXP start, SEXP end, SEXP width)
{
	int ans_len, use_start_as_is, use_width_as_is,
	    i, solved_start, solved_width,
		start_value, end_value, width_value;
	SEXP ans, ans_start, ans_width,
		check_start_NA, check_width_NA;

	//These functions might not only be called by constructor?
	if (!(IS_INTEGER(start) && IS_INTEGER(end) && IS_INTEGER(width)))
		error("the supplied 'start', 'end', and 'width', "
                      "must be integer vectors");
	ans_len = LENGTH(start);
	if (LENGTH(end) != ans_len || LENGTH(width) != ans_len)
		error("'start', 'end', and 'width' must have the same length");

	/*Prepare the expression to check any NA values in start and width
	  The function anyNA can be more efficient than a loop in C.*/
	PROTECT(check_start_NA = lang2(install("anyNA"), start));
	PROTECT(check_width_NA = lang2(install("anyNA"), width));

	use_start_as_is = GET_DIM(start) == R_NilValue &&
		GET_NAMES(start) == R_NilValue &&
		!asLogical(R_tryEval(check_start_NA, R_GlobalEnv, NULL));
	use_width_as_is = GET_DIM(width) == R_NilValue &&
		GET_NAMES(width) == R_NilValue&&
		!asLogical(R_tryEval(check_width_NA, R_GlobalEnv, NULL));

	ans_start = start;
	ans_width = width;
	if (!(use_start_as_is && use_width_as_is)) {
		/* Allocate and populate 'ans_start' and/or 'ans_width'. 
		   Call the duplicate function to duplicate an object for
		   an ALTREP might has a more efficient way to duplicate itself*/
		if (!use_start_as_is)
			PROTECT(ans_start = duplicate(start));
		if (!use_width_as_is)
			PROTECT(ans_width = duplicate(width));
		
		for (i = 0; i < ans_len; i++) {
			start_value = INTEGER_ELT(start, i);
			end_value = INTEGER_ELT(end, i);
			width_value = INTEGER_ELT(width, i);
			if (solve_range(start_value, end_value, width_value,
				&solved_start, &solved_width) != 0)
				error("In range %d: %s.", i + 1, errmsg_buf);
			/*The if statement is used to avoid unnecessary assignment*/
			if (!use_start_as_is)
				SET_INTEGER_ELT(ans_start, i, solved_start);
			if (!use_width_as_is)
				SET_INTEGER_ELT(ans_width, i, solved_width);
		}
	}
	PROTECT(ans = _new_IRanges("IRanges", ans_start, ans_width,
					      R_NilValue));
	UNPROTECT(3 + !use_start_as_is + !use_width_as_is);
	return ans;
}


/****************************************************************************
 * solve_user_SEW()
 */

static int translate_negative_coord0;
static int nonnarrowing_is_OK;

static int translate_negative_startorend(int refwidth, int startorend)
{
	if (startorend < 0)
		startorend += refwidth + 1;
	return startorend;
}

static int check_start(int refwidth, int start, const char *what)
{
	if (nonnarrowing_is_OK)
		return 0;
	if (start < 1) {
		snprintf(errmsg_buf, sizeof(errmsg_buf),
			 "'allow.nonnarrowing' is FALSE and the %s start "
			 "(%d) is < 1", what, start);
		return -1;
	}
	if (start > refwidth + 1) {
		snprintf(errmsg_buf, sizeof(errmsg_buf),
			 "'allow.nonnarrowing' is FALSE and the %s start "
			 "(%d) is > refwidth + 1", what, start);
		return -1;
	}
	return 0;
}

static int check_end(int refwidth, int end, const char *what)
{
	if (nonnarrowing_is_OK)
		return 0;
	if (end < 0) {
		snprintf(errmsg_buf, sizeof(errmsg_buf),
			 "'allow.nonnarrowing' is FALSE and the %s end "
			 "(%d) is < 0", what, end);
		return -1;
	}
	if (end > refwidth) {
		snprintf(errmsg_buf, sizeof(errmsg_buf),
			 "'allow.nonnarrowing' is FALSE and the %s end "
			 "(%d) is > refwidth", what, end);
		return -1;
	}
	return 0;
}

static int solve_user_SEW_row(int refwidth, int start, int end, int width,
		int *solved_start, int *solved_width)
{
	if (refwidth == NA_INTEGER || refwidth < 0) {
		snprintf(errmsg_buf, sizeof(errmsg_buf),
			 "negative values or NAs are not allowed in 'refwidths'");
		return -1;
	}
	if (start != NA_INTEGER) {
		if (translate_negative_coord0)
			start = translate_negative_startorend(refwidth, start);
		if (check_start(refwidth, start, "supplied") != 0)
			return -1;
	}
	if (end != NA_INTEGER) {
		if (translate_negative_coord0)
			end = translate_negative_startorend(refwidth, end);
		if (check_end(refwidth, end, "supplied") != 0)
			return -1;
	}
	if (width == NA_INTEGER) {
		if (start == NA_INTEGER)
			start = 1;
		if (end == NA_INTEGER)
			end = refwidth;
		width = end - start + 1;
		if (width < 0) {
			snprintf(errmsg_buf, sizeof(errmsg_buf),
				 "the supplied start/end lead to a "
				 "negative width");
			return -1;
		}
	} else if (width < 0) {
		snprintf(errmsg_buf, sizeof(errmsg_buf),
			 "negative values are not allowed in 'width'");
		return -1;
	} else if ((start == NA_INTEGER) == (end == NA_INTEGER)) {
		snprintf(errmsg_buf, sizeof(errmsg_buf),
			 "either the supplied start or the supplied end "
			 "(but not both) must be NA when the supplied width "
			 "is not NA");
		return -1;
	} else {
		// Either 'start' or 'end' is NA
		if (start == NA_INTEGER) {
			start = end - width + 1;
			if (check_start(refwidth, start, "solved") != 0)
				return -1;
		} else {
			end = start + width - 1;
			if (check_end(refwidth, end, "solved") != 0)
				return -1;
		}
	}
	*solved_start = start;
	*solved_width = width;
	return 0;
}

/*
 * --- .Call ENTRY POINT ---
 */
SEXP solve_user_SEW(SEXP refwidths, SEXP start, SEXP end, SEXP width,
		SEXP translate_negative_coord, SEXP allow_nonnarrowing)
{
	SEXP ans, ans_start, ans_width,
		check_start_NA, check_width_NA;
	int ans_len, i0, i1, i2, i3,
		refwidths_value, start_value, width_value, end_value,
		use_start_as_is, use_width_as_is,
		solved_start, solved_width;

	translate_negative_coord0 = LOGICAL(translate_negative_coord)[0];
	nonnarrowing_is_OK = LOGICAL(allow_nonnarrowing)[0];
	ans_len = LENGTH(refwidths);

	/*Prepare the expression to check any NA values in start and width
	  The function anyNA can be more efficient than a loop in C.*/
	PROTECT(check_start_NA = lang2(install("anyNA"), start));
	PROTECT(check_width_NA = lang2(install("anyNA"), width));

	/*Please see comments in solve_user_SEW0
	  The length should alse be check since it is not checked in R*/
	use_start_as_is = GET_DIM(start) == R_NilValue &&
		GET_NAMES(start) == R_NilValue &&
		!asLogical(R_tryEval(check_start_NA, R_GlobalEnv, NULL))&&
		LENGTH(start) == ans_len
		;
	use_width_as_is = GET_DIM(width) == R_NilValue &&
		GET_NAMES(width) == R_NilValue &&
		!asLogical(R_tryEval(check_width_NA, R_GlobalEnv, NULL))&&
		LENGTH(width) == ans_len;

	/*
	If user's input cannot be used `as-is`, depending on the vector's length:
	  1. If the length is the same as ans_len, the vector will be duplicated
	  2. If the length is different from ans_len, a new vector will be created.
	*/
	if (!use_start_as_is) {
		if (LENGTH(start) == ans_len) 
			PROTECT(ans_start = duplicate(start));
		else
			PROTECT(ans_start = NEW_INTEGER(ans_len));
	}
	else {
		ans_start = start;
	}

	if (!use_width_as_is) {
		if (LENGTH(width) == ans_len)
			PROTECT(ans_width = duplicate(width));
		else
			PROTECT(ans_width = NEW_INTEGER(ans_len));
	}
	else {
		ans_width = width;
	}


	for (i0 = i1 = i2 = i3 = 0; i0 < ans_len; i0++, i1++, i2++, i3++) {
		/* recycling */
		if (i1 >= LENGTH(start)) i1 = 0;
		if (i2 >= LENGTH(end)) i2 = 0;
		if (i3 >= LENGTH(width)) i3 = 0;

		refwidths_value = INTEGER_ELT(refwidths, i0);
		start_value = INTEGER_ELT(start, i1);
		end_value = INTEGER_ELT(end, i2);
		width_value = INTEGER_ELT(width, i3);

		if (solve_user_SEW_row(INTEGER(refwidths)[i0],
				       INTEGER(start)[i1],
				       INTEGER(end)[i2],
				       INTEGER(width)[i3],
			           solved_start,
			           solved_width
				       ) != 0)
		{
			UNPROTECT(2 + use_start_as_is + use_width_as_is);
			error("solving row %d: %s", i0 + 1, errmsg_buf);
		}
		if (!use_start_as_is)
			SET_INTEGER_ELT(ans_start, i0, solved_start);
		if (!use_width_as_is)
			SET_INTEGER_ELT(ans_width, i0, solved_width);
	}
	PROTECT(ans = _new_IRanges("IRanges", ans_start, ans_width, R_NilValue));
	UNPROTECT(3 + use_start_as_is + use_width_as_is);
	return ans;
}

File: R/IRanges-constructor.R
### =========================================================================
### The IRanges constructor
### -------------------------------------------------------------------------
###


### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
### Two low-level helpers
###
### Input can contain NAs. Output must be an unnamed integer vector.
###

.start_as_unnamed_integer <- function(start, what="a start")
{
    if (is.integer(start))
        return(unname(start))
    old_warn <- getOption("warn")
    options(warn=2L)
    on.exit(options(warn=old_warn))
    start <- try(as.integer(start), silent=TRUE)
    if (inherits(start, "try-error"))
        stop(wmsg("each range must have ", what, " that ",
                  "is < 2^31 and > - 2^31"))
    start
}

.width_as_unnamed_integer <- function(width, msg="a non-negative width")
{
    if (any(width < 0, na.rm=TRUE))
        stop(wmsg("each range must have ", msg))
    if (is.integer(width))
        return(unname(width))
    old_warn <- getOption("warn")
    options(warn=2L)
    on.exit(options(warn=old_warn))
    width <- try(as.integer(width), silent=TRUE)
    if (inherits(width, "try-error"))
        stop(wmsg("each range must have a width that is < 2^31"))
    width
}


### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
### Low-level IRanges constructors
###

.new_empty_IRanges <- function() new2("IRanges", check=FALSE)

.new_IRanges_from_start_end <- function(start, end)
{
    if (!is.numeric(start) || !is.numeric(end))
        stop(wmsg("'start' and 'end' must be numeric vectors"))
    if (anyNA(start) || anyNA(end))
        stop(wmsg("'start' and 'end' cannot contain NAs"))
    if (length(start) == 0L || length(end) == 0L)
        return(.new_empty_IRanges())
    start <- .start_as_unnamed_integer(start)
    end   <- .start_as_unnamed_integer(end, what="an end")
    ## We want to perform this operation in "double" space rather
    ## than in "integer" space so we use 1.0 instead of 1L.
    width <- 1.0 + end - start
    width <- .width_as_unnamed_integer(width,
                 msg="an end that is greater or equal to its start minus one")
    start <- S4Vectors:::recycleVector(start, length(width))
    new2("IRanges", start=start, width=width, check=FALSE)
}

.new_IRanges_from_start_width <- function(start, width)
{
    if (!is.numeric(start) || !is.numeric(width))
        stop(wmsg("'start' and 'width' must be numeric vectors"))
    if (anyNA(start) || anyNA(width))
        stop(wmsg("'start' and 'width' cannot contain NAs"))
    if (length(start) == 0L || length(width) == 0L)
        return(.new_empty_IRanges())
    start <- .start_as_unnamed_integer(start)
    width <- .width_as_unnamed_integer(width)
    ## We want to perform this operation in "double" space rather
    ## than in "integer" space so we use -1.0 instead of -1L.
    end <- -1.0 + start + width
    end <- .start_as_unnamed_integer(end, what="an end")
    start <- S4Vectors:::recycleVector(start, length(end))
    width <- S4Vectors:::recycleVector(width, length(end))
    new2("IRanges", start=start, width=width, check=FALSE)
}

.new_IRanges_from_end_width <- function(end, width)
{
    if (!is.numeric(end) || !is.numeric(width))
        stop(wmsg("'end' and 'width' must be numeric vectors"))
    if (anyNA(end) || anyNA(width))
        stop(wmsg("'end' and 'width' cannot contain NAs"))
    if (length(end) == 0L || length(width) == 0L)
        return(.new_empty_IRanges())
    end   <- .start_as_unnamed_integer(end, what="an end")
    width <- .width_as_unnamed_integer(width)
    ## We want to perform this operation in "double" space rather
    ## than in "integer" space so we use 1.0 instead of 1L.
    start <- 1.0 + end - width
    start <- .start_as_unnamed_integer(start)
    start <- suppressWarnings(as.integer(start))
    width <- S4Vectors:::recycleVector(width, length(start))
    new2("IRanges", start=start, width=width, check=FALSE)
}

.is_numeric_or_NAs <- function(x)
{
    is.numeric(x) || is.logical(x) && all(is.na(x))
}

.solve_start_width_end <- function(start, end, width)
{
    if (!.is_numeric_or_NAs(start)
     || !.is_numeric_or_NAs(end)
     || !.is_numeric_or_NAs(width))
        stop(wmsg("'start', 'end', and 'width', must be numeric vectors"))
    L1 <- length(start)
    L2 <- length(end)
    L3 <- length(width)
    if (min(L1, L2, L3) == 0L)
        return(.new_empty_IRanges())
    if (is.logical(start)) {
        start <- as.integer(start)
    } else {
        start <- .start_as_unnamed_integer(start)
    }
    if (is.logical(end)) {
        end <- as.integer(end)
    } else {
        end <- .start_as_unnamed_integer(end, what="an end")
    }
    if (is.logical(width)) {
        width <- as.integer(width)
    } else {
        width <- .width_as_unnamed_integer(width)
    }
    ans_len <- max(L1, L2, L3)
    start <- S4Vectors:::recycleVector(start, ans_len)
    end   <- S4Vectors:::recycleVector(end, ans_len)
    width <- S4Vectors:::recycleVector(width, ans_len)
    .Call2("C_solve_start_width_end", start, end, width, PACKAGE="IRanges")
}

.new_IRanges <- function(start=NULL, end=NULL, width=NULL)
{
    start_is_null <- is.null(start)
    end_is_null <- is.null(end)
    width_is_null <- is.null(width)
    nb_of_nulls <- sum(start_is_null, end_is_null, width_is_null)
    if (nb_of_nulls == 3L)
        return(.new_empty_IRanges())
    if (nb_of_nulls == 2L)
        stop(wmsg("at least two of the 'start', 'end', and 'width' ",
                  "arguments must be supplied"))
    if (width_is_null)
        return(.new_IRanges_from_start_end(start, end))
    if (end_is_null)
        return(.new_IRanges_from_start_width(start, width))
    if (start_is_null)
        return(.new_IRanges_from_end_width(end, width))
    .solve_start_width_end(start, end, width)
}


### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
### High-level IRanges constructor
###

IRanges <- function(start=NULL, end=NULL, width=NULL, names=NULL, ...)
{
    mcols <- DataFrame(..., check.names=FALSE)

    if (!is.null(start) && is.null(end) && is.null(width)) {
        ans <- as(start, "IRanges")
    } else {
        ans <- .new_IRanges(start=start, end=end, width=width)
    }

    if (!is.null(names))
        names(ans) <- names
    if (length(mcols) != 0L)
        mcols(ans) <- mcols
    ans
}


### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
### The SEW interface: start=NA/end=NA/width=NA
###
### Some of the functions that support the SEW interface: narrow(),
### XVector::subseq(), XVector::xvcopy(), Biostrings::BStringSet() (and
### family), BSgenome::getSeq(), etc...
###

.normargSEW <- function(x, argname)
{
    if (!S4Vectors:::isNumericOrNAs(x))
        stop("'", argname, "' must be a vector of integers")
    if (!is.integer(x))
        x <- as.integer(x)
    x
}

### Use of 'rep.refwidths=TRUE' is supported only when 'refwidths' is of
### length 1.
### If 'rep.refwidths=FALSE' (the default) then 'start', 'end' and 'width'
### are recycled to 'length(refwidths)' (it's an error if one of them is
### longer than 'refwidths'). Otherwise, 'refwidths' is replicated L times
### where L is the length of the longest of 'start', 'end' and 'width'.
### The returned value is an IRanges object of the same length as 'refwidths'
### (after replication if 'rep.refwidths=TRUE').
solveUserSEW <- function(refwidths, start=NA, end=NA, width=NA,
                         rep.refwidths=FALSE,
                         translate.negative.coord=TRUE,
                         allow.nonnarrowing=FALSE)
{
    if (!is.numeric(refwidths))
        stop("'refwidths' must be a vector of integers")
    if (!is.integer(refwidths))
        refwidths <- as.integer(refwidths)

    start <- .normargSEW(start, "start")
    end <- .normargSEW(end, "end")
    width <- .normargSEW(width, "width")
    ## From here, 'refwidths', 'start', 'end' and 'width' are guaranteed to be
    ## integer vectors. NAs in 'start', 'end' and 'width' are OK but not in
    ## 'refwidths' so this should be checked at the C level.

    if (!isTRUEorFALSE(rep.refwidths))
        stop("'rep.refwidths' must be TRUE or FALSE")

    if (!isTRUEorFALSE(translate.negative.coord))
        stop("'translate.negative.coord' must be TRUE or FALSE")
    if (!isTRUEorFALSE(allow.nonnarrowing))
        stop("'allow.nonnarrowing' must be TRUE or FALSE")

    Lsew <- c(length(start), length(end), length(width))
    maxLsew <- max(Lsew)
    minLsew <- min(Lsew)
    if (minLsew == 0L && maxLsew > 1L)
        stop("'start', 'end' and 'width' cannot mix zero-length ",
             "and longer-than-one vectors")

    ## Check 'start', 'end', and 'width' *without* recycling them. Recycling
    ## is done at the C level.
    if (rep.refwidths) {
        if (length(refwidths) != 1L)
            stop("'rep.refwidths=TRUE' can be used only when 'refwidths' ",
                 "is of length 1")
        ## 'ans_len' is the length of the longest of 'start', 'end'
        ## and 'width'.
        if (minLsew == 0L) {
            ans_len <- 0L
        } else {
            ans_len <- maxLsew
        }
        refwidths <- rep.int(refwidths, ans_len)
    } else {
        ans_len <- length(refwidths)
        if (ans_len == 0L) {
            if (maxLsew > 1L)
                stop("'start', 'end' or 'width' is longer than 'refwidths'")
        } else {
            if (minLsew == 0L)
                stop("cannot recycle empty 'start', 'end' or 'width'")
            if (maxLsew > ans_len)
                stop("'start', 'end' or 'width' is longer than 'refwidths'")
        }
    }

    .Call2("C_solve_user_SEW",
           refwidths, start, end, width,
           translate.negative.coord, allow.nonnarrowing,
           PACKAGE="IRanges")
}

### Returns an IRanges instance of length 1. Not exported.
solveUserSEWForSingleSeq <- function(x_length, start=NA, end=NA, width=NA)
{
    solved_SEW <-
      try(solveUserSEW(x_length, start=start, end=end, width=width),
          silent = TRUE)
    if (is(solved_SEW, "try-error"))
        stop("Invalid sequence coordinates.\n",
             "  Please make sure the supplied 'start', 'end' and 'width' arguments\n",
             "  are defining a region that is within the limits of the sequence.")
    solved_SEW
}

